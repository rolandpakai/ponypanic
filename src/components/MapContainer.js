import { Fragment, useEffect, useState, useContext } from 'react';

import { NewGameContext } from '../contexts/NewGameContext';
import { apiStoryBegin, apiMapResource, apiMapState, apiApproveHeroTurn, apiPlaythroughState, apiResetLevel, apiNextLevel } from '../api/api';
import { GAME_MODE, PLAYER_TOKEN, FIELD_TYPE, MAP_STATUS } from '../utils/constants';
import { arrayToMap, obstacleMapToArray, getImageSize } from '../utils/util';
import Canvas from './Canvas';
import PopupDialog from './PopupDialog';
import CustomButton from './CustomButton';

const MapContainer = () => {
  const [gameMode, setGameMode] = useState(GAME_MODE.STORY);
  const {newGame, setNewGame} = useContext(NewGameContext);
  const [canvas, setCanvas] = useState({});
  const [storyPlaythroughToken, setStoryPlaythroughToken] = useState('');
  const [heroTurn, setHeroTurn] = useState({});
  const [dialogProps, setDialogProps] = useState({open:false});

  const updateHeroTurn = (heroTurn) => {
    setHeroTurn(heroTurn)
  }

  const getCanvasData = (mapResource, mapState, currentLevel) => {
    const { mapId, compressedObstacles: {coordinateMap} } = mapResource;
    const { map, heroes } = mapState;

    const size = getImageSize(map.width);
    const obstaclesList = obstacleMapToArray(coordinateMap);
    const heroesList = arrayToMap(heroes, FIELD_TYPE.HERO);
    const treasures = arrayToMap(map.treasures, FIELD_TYPE.TREASURE);
    const obstacles = arrayToMap(obstaclesList, FIELD_TYPE.OBSTACLE);
    const enemies = arrayToMap(map.enemies, FIELD_TYPE.ENEMY);
    const bullets = arrayToMap(map.bullets, FIELD_TYPE.BULLET);

    return {
      mapId: mapId,
      width: map.width,
      height: map.height,
      fieldSize: size,
      heroes: heroesList,
      enemies: enemies,
      bullets: bullets,
      treasures: treasures,
      obstacles: obstacles,
      collected: {},
      currentLevel: currentLevel,
      gameMode: gameMode,
      updateHeroTurn: updateHeroTurn,
    };

  }

  const resetLevel = async () => {
    const { playthroughState: {currentLevel, isCurrentLevelFinished, currentMapStatus} } = await apiResetLevel(storyPlaythroughToken);
    const mapResource= await apiMapResource(storyPlaythroughToken);
    const mapState = await apiMapState(storyPlaythroughToken);
    const newCanvas = getCanvasData(mapResource, mapState, currentLevel)

    setDialogProps({...dialogProps, open: false});
    setCanvas(newCanvas);
  }

  const nextLevel = async () => {
    const { playthroughState: {currentLevel, isCurrentLevelFinished, currentMapStatus} } = await apiNextLevel(storyPlaythroughToken);
    const mapResource = await apiMapResource(storyPlaythroughToken);
    const mapState = await apiMapState(storyPlaythroughToken);
    const newCanvas = getCanvasData(mapResource, mapState, currentLevel)

    setDialogProps({...dialogProps, open: false});
    setCanvas(newCanvas);
  }

  const endHandler = () => {
    console.log('endHandler');
    setNewGame(false);
  };

  const continueHandler = () => {
    console.log('continueHandler')
    resetLevel(storyPlaythroughToken);
  };

  const nextHandler = () => {
    console.log('nextHandler')
    nextLevel();
  };

  const dialogPropsStatus = {
    'LOST': {
      open: true,
      dialogContentText: 'GAME OVER',
      buttonOkText: 'CONTINUE',
      buttonCancelText: 'END', 
      okHandler: continueHandler,
      cancelHandler: endHandler,
    },
    'WON': {
      open: true,
      dialogContentText: 'CONGRATULATION',
      buttonOkText: 'NEXT LEVEL',
      buttonCancelText: 'END', 
      okHandler: nextHandler,
      cancelHandler: endHandler,
    },
  };

  const nextTurn = async (heroTurn) => {
    const { didTickHappen, message, tickLogs } = await apiApproveHeroTurn(storyPlaythroughToken, heroTurn)

    if(didTickHappen) {
      const { map, heroes } = await apiMapState(storyPlaythroughToken);

      if(map.isGameOver) {
        const {currentLevel, currentMapStatus, isCurrentLevelFinished} = await apiPlaythroughState(storyPlaythroughToken);
        
        if(map.status === MAP_STATUS.WON) {
          setDialogProps(dialogPropsStatus[MAP_STATUS.WON]);
        } else if(map.status === MAP_STATUS.LOST) {
          setDialogProps(dialogPropsStatus[MAP_STATUS.LOST]);
        }
      } else {
        //MAP_STATUS.CRATED || MAP_STATUS.PLAYING
        const updates = {
          heroes: arrayToMap(heroes, FIELD_TYPE.HERO, heroTurn),
          enemies: arrayToMap(map.enemies, FIELD_TYPE.ENEMY),
          bullets: arrayToMap(map.bullets, FIELD_TYPE.BULLET),
        }
        const treasures = arrayToMap(map.treasures, FIELD_TYPE.TREASURE);
        const collectedList = Object.values(treasures).filter((treasure)=>treasure.collectedByHeroId!=null);
        
        if(collectedList.length > 0) {
          const collected = arrayToMap(collectedList, FIELD_TYPE.COLLECTED_TREASURE);
          updates.collected = collected;
        }

        setCanvas({
          ...canvas,
          ...updates
        });
      }
    }
  }

  const nextTurnHandle = () => {
    heroTurn.storyPlaythroughToken = storyPlaythroughToken;

    nextTurn(heroTurn);
  }

  useEffect(() => {

    const fetchData = async () => {
      const {storyPlaythroughToken, playthroughState: {currentLevel, isCurrentLevelFinished, currentMapStatus} } = await apiStoryBegin(PLAYER_TOKEN);
      const mapResource = await apiMapResource(storyPlaythroughToken);
      const mapState = await apiMapState(storyPlaythroughToken);
      const canvas = getCanvasData(mapResource, mapState, currentLevel)

      setStoryPlaythroughToken(storyPlaythroughToken);
      setCanvas(canvas);
    }

    fetchData();
  }, []);

  return (
    <Fragment>   
      <main className="main">
        <div className="main-container">
          <div className="sub-container">
            <div className="panel-left">
                <Canvas 
                  {...canvas}
                />
            </div>
          </div>
          <CustomButton onClick={nextTurnHandle} title={"HERO TURN"}/>
        </div>
      </main>
      <PopupDialog {...dialogProps} />
    </Fragment>
  )
}

export default MapContainer;