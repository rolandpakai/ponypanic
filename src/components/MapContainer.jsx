import React, { useContext, useEffect, useState } from "react";
import Box from "@mui/material/Box";
import Container from "@mui/material/Container";
import Stack from "@mui/material/Stack";
import CircularProgress from "@mui/material/CircularProgress";

import {
  apiApproveHeroTurn,
  apiMapResource,
  apiMapState,
  apiNextLevel,
  apiPlaythroughState,
  apiResetLevel,
  apiStoryBegin,
} from "../api/api";
import {
  arrayToMap,
  getImageSize,
  mapToArray,
  existsItemInLocalStorage,
  getItemFromLocalStorage,
} from "../utils/util";
import {
  FIELD_TYPE,
  GAME_MODE,
  GAME_STATE,
  LOADING_TIMEOUT,
  PLAYER_TOKEN,
  LOCAL_STORAGE_STATE_NAME,
} from "../utils/constants";
import Canvas from "./Canvas";
import PopupDialog from "./PopupDialog";
import Button from "./Button";
import { GameStateContext } from "../contexts/GameStateContext";
import { GameModeContext } from "../contexts/GameModeContext";

const MapContainer = () => {
  const { updateGameMode } = useContext(GameModeContext);
  const { gameState, updateGameState } = useContext(GameStateContext);
  const [canvas, setCanvas] = useState({});
  const [storyToken, setStoryToken] = useState("");
  const [heroTurn, setHeroTurn] = useState({});
  const [dialogProps, setDialogProps] = useState({ open: false });
  const [loading, setLoading] = useState(false);
  const [loadingTurn, setLoadingTurn] = useState(false);

  const updateCanvas = (token, newCanvas) => {
    const newGameState = {
      ...gameState,
      value: { storyToken: token, canvas: newCanvas },
    };

    updateGameState(newGameState);
    setCanvas(newCanvas);
  };

  const updateHeroTurn = (nextHeroTurn) => {
    setHeroTurn(nextHeroTurn);
  };

  const getCanvasData = (mapResource, mapState, currentLevel) => {
    const {
      mapId,
      compressedObstacles: { coordinateMap },
    } = mapResource;

    const { map, heroes } = mapState;

    const size = getImageSize(map.width);
    const obstaclesList = mapToArray(coordinateMap);
    const heroesList = arrayToMap(heroes, FIELD_TYPE.HERO);
    const treasures = arrayToMap(map.treasures, FIELD_TYPE.TREASURE);
    const obstacles = arrayToMap(obstaclesList, FIELD_TYPE.OBSTACLE);
    const enemies = arrayToMap(map.enemies, FIELD_TYPE.ENEMY);
    const bullets = arrayToMap(map.bullets, FIELD_TYPE.BULLET);

    return {
      mapId,
      width: map.width,
      height: map.height,
      fieldSize: size,
      heroes: heroesList,
      enemies,
      bullets,
      treasures,
      obstacles,
      collected: {},
      currentLevel,
      elapsedTickCount: map.elapsedTickCount,
    };
  };

  const setCanvasData = async (token, currentLevel) => {
    const mapResource = await apiMapResource(token);
    const mapState = await apiMapState(token);
    const newCanvas = getCanvasData(mapResource, mapState, currentLevel);

    updateCanvas(token, newCanvas);
  };

  const resetLevel = async (token) => {
    const {
      playthroughState: { currentLevel },
    } = await apiResetLevel(token);

    setCanvasData(token, currentLevel);
    setDialogProps({ ...dialogProps, open: false });
  };

  const nextLevel = async (token) => {
    const {
      playthroughState: { currentLevel },
    } = await apiNextLevel(token);

    setCanvasData(token, currentLevel);
    setDialogProps({ ...dialogProps, open: false });
  };

  const endLevel = async (token) => {
    const {
      playthroughState: { currentLevel },
    } = await apiNextLevel(token);

    const mapResource = await apiMapResource(token);
    const mapState = await apiMapState(token);
    const newCanvas = getCanvasData(mapResource, mapState, currentLevel);

    const newGameState = {
      state: GAME_STATE.END,
      value: { storyToken: token, canvas: newCanvas },
    };

    updateGameState(newGameState);
    updateGameMode(GAME_MODE.UNDEFINED);
  };

  const endHandler = () => {
    endLevel(storyToken);
  };

  const continueHandler = () => {
    resetLevel(storyToken);
  };

  const nextHandler = () => {
    nextLevel(storyToken);
  };

  const dialogPropsStatus = {
    LOST: {
      open: true,
      dialogContentText: "GAME OVER",
      buttonOkText: "CONTINUE",
      buttonCancelText: "END",
      okHandler: continueHandler,
      cancelHandler: endHandler,
    },
    WON: {
      open: true,
      dialogContentText: "CONGRATULATION",
      buttonOkText: "NEXT LEVEL",
      buttonCancelText: "END",
      okHandler: nextHandler,
      cancelHandler: endHandler,
    },
  };

  const getUpdatedData = (map, heroes, turn) => {
    const updates = {
      heroes: arrayToMap(heroes, FIELD_TYPE.HERO, turn),
      enemies: arrayToMap(map.enemies, FIELD_TYPE.ENEMY),
      bullets: arrayToMap(map.bullets, FIELD_TYPE.BULLET),
      elapsedTickCount: map.elapsedTickCount,
    };

    const treasures = arrayToMap(map.treasures, FIELD_TYPE.TREASURE);

    const collectedList = Object.values(treasures).filter(
      (treasure) => treasure.collectedByHeroId != null
    );

    if (collectedList.length > 0) {
      const collected = arrayToMap(
        collectedList,
        FIELD_TYPE.COLLECTED_TREASURE
      );

      updates.collected = collected;
    }

    return updates;
  };

  const showDialog = (mapStatus) => {
    setDialogProps(dialogPropsStatus[mapStatus]);
  };

  const gameOver = async (mapStatus) => {
    await apiPlaythroughState(storyToken);
    showDialog(mapStatus);
  };

  const didTickHappened = async (turn) => {
    const { map, heroes } = await apiMapState(storyToken);
    const updates = getUpdatedData(map, heroes, turn);

    const newCanvas = {
      ...canvas,
      ...updates,
    };

    updateCanvas(storyToken, newCanvas);

    setTimeout(() => {
      setLoadingTurn(false);
    }, LOADING_TIMEOUT);

    if (map.isGameOver) {
      gameOver(map.status);
    }
  };

  const nextTurn = async (turn) => {
    setLoadingTurn(true);

    const { didTickHappen } = await apiApproveHeroTurn(storyToken, turn);

    if (didTickHappen) {
      didTickHappened(turn);
    }
  };

  const nextTurnHandle = () => {
    nextTurn(heroTurn);
  };

  const fetchData = async () => {
    const {
      storyPlaythroughToken,
      playthroughState: { currentLevel },
    } = await apiStoryBegin(PLAYER_TOKEN);

    setCanvasData(storyPlaythroughToken, currentLevel);
    setStoryToken(storyPlaythroughToken);
    setDialogProps({ ...dialogProps, open: false });

    setTimeout(() => {
      setLoading(false);
    }, LOADING_TIMEOUT);
  };

  const loadData = () => {
    if (existsItemInLocalStorage(LOCAL_STORAGE_STATE_NAME)) {
      const storedState = getItemFromLocalStorage(LOCAL_STORAGE_STATE_NAME);

      updateCanvas(storedState.storyToken, storedState.canvas);
      setStoryToken(storedState.storyToken);
      setDialogProps({ ...dialogProps, open: false });

      setTimeout(() => {
        setLoading(false);
      }, LOADING_TIMEOUT);
    }
  };

  useEffect(() => {
    setLoading(true);

    if (gameState.state === GAME_STATE.NEW) {
      fetchData();
    }

    if (gameState.state === GAME_STATE.CONTINUE) {
      loadData();
    }
  }, []);

  return (
    <>
      <Box>
        <Container maxWidth="md">
          <Stack
            direction="row"
            justifyContent="center"
            maxWidth="xs"
            sx={{
              padding: "2rem",
              minHeight: "520px",
              alignItems: "center",
            }}
          >
            {loading ? (
              <CircularProgress
                color="inherit"
                size={150}
                sx={{ color: "var(--progress-color)" }}
              />
            ) : (
              <Box
                sx={{
                  border: "5px solid #3a3b3b",
                }}
              >
                <Canvas canvas={canvas} updateHeroTurn={updateHeroTurn} />
              </Box>
            )}
          </Stack>
        </Container>
        <Container maxWidth="xs">
          <Stack direction="row" justifyContent="center" maxWidth="xs">
            <Button loading={loading || loadingTurn} onClick={nextTurnHandle}>
              HERO TURN
            </Button>
          </Stack>
        </Container>
      </Box>
      <PopupDialog
        open={dialogProps.open}
        dialogContentText={dialogProps.dialogContentText}
        buttonOkText={dialogProps.buttonOkText}
        buttonCancelText={dialogProps.buttonCancelText}
        okHandler={dialogProps.okHandler}
        cancelHandler={dialogProps.cancelHandler}
      />
    </>
  );
};

export default MapContainer;
