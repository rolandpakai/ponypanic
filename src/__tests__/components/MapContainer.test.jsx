import React from "react";
import renderer from "react-test-renderer";
import MapContainer from "../../components/MapContainer";

describe("MapContainer Component", () => {
  test("MapContainer renders", () => {
    const container = renderer.create(<MapContainer />).toJSON();

    expect(container).toMatchSnapshot();
  });
});
