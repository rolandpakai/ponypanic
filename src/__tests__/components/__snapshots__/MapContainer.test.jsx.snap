// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapContainer Component MapContainer renders 1`] = `
<div
  className="MuiBox-root css-0"
>
  <div
    className="MuiContainer-root MuiContainer-maxWidthMd css-xdmu94-MuiContainer-root"
  >
    <div
      className="css-1av9md8-MuiStack-root"
    >
      <div
        className="MuiBox-root css-1as3uml"
      >
        <div
          className="MuiBox-root css-gaqaxa"
          id="id-undefined"
        />
      </div>
    </div>
  </div>
  <div
    className="MuiContainer-root MuiContainer-maxWidthXs css-nzinn5-MuiContainer-root"
  >
    <div
      className="css-vankme-MuiStack-root"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth css-dat04h-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
        disabled={false}
        id=":r0:"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        HERO TURN
      </button>
    </div>
  </div>
</div>
`;
